src/App.jsx

```diff
@@ -0,0 +1,165 @@
+
+function App() {
+  const [cardNumber, setCardNumber] = useState(["", "", "", ""]);
+  const [expireDate, setExpireDate] = useState(["", ""]);
+  const [holderName, setHolderName] = useState("");
+  const [securityCode, setSecurityCode] = useState("");
+  const [password, setPassword] = useState(["", ""]);
+  const [canProceed, setCanProceed] = useState(false);
+
+  const handleCardNumberUpdate = ({ target: { value } }, order) => {
+    if (
+      !Number.isInteger(Number(value)) ||
+      value.length > CARD_INFO_RULES.NUMBER_UNIT_LENGTH
+    )
+      return;
+
+    setCardNumber((prevValue) => {
+      const newValue = [...prevValue];
+      newValue[order] = value;
+
+      return newValue;
+    });
+  };
+
+  const handleExpireDateUpdate = ({ target: { value } }, order) => {
+    const parsedValue =
+      value.startsWith("0") && value.length !== 1 ? value.slice(1) : value;
+
+    if (!/^\d{0,2}$/.test(parsedValue)) return;
+
+    if (
+      order === 0 &&
+      value !== "0" &&
+      value !== "" &&
+      (Number(parsedValue) > 12 || Number(parsedValue) < 1)
+    ) {
+      return;
+    }
+
+    setExpireDate((prevValue) => {
+      const newValue = [...prevValue];
+      newValue[order] =
+        parsedValue.length === 1 && Number(parsedValue) !== 0
+          ? `0${parsedValue}`
+          : parsedValue;
+
+      return newValue;
+    });
+  };
+
+  const handleHolderNameUpdate = ({ target: { value } }) => {
+    if (
+      !/^[a-z]*$/i.test(value) ||
+      value.length > CARD_INFO_RULES.HOLDER_NAME_MAX_LENGTH
+    )
+      return;
+
+    setHolderName(value.toUpperCase());
+  };
+
+  const handleSecurityCodeUpdate = ({ target: { value } }) => {
+    if (
+      !Number.isInteger(Number(value)) ||
+      value.length > CARD_INFO_RULES.SECURITY_CODE_LENGTH
+    )
+      return;
+
+    setSecurityCode(value);
+  };
+
+  const handlePasswordUpdate = ({ target: { value } }, order) => {
+    if (!Number.isInteger(Number(value)) || value.length > 1) return;
+
+    setPassword((prevValue) => {
+      const newValue = [...prevValue];
+      newValue[order] = value;
+
+      return newValue;
+    });
+  };
+
+  const handleCardInfoSubmit = () => {
+    alert(CARD_REGISTER_SUCCESS_MESSAGE);
+  };
+
+  const isValidCardInfo = useCallback(() => {
+    const {
+      NUMBER_UNIT_COUNT,
+      NUMBER_UNIT_LENGTH,
+      EXPIRE_DATE_LENGTH,
+      SECURITY_CODE_LENGTH,
+      PASSWORD_LENGTH,
+    } = CARD_INFO_RULES;
+
+    return (
+      cardNumber.join("").length === NUMBER_UNIT_COUNT * NUMBER_UNIT_LENGTH &&
+      expireDate.join("").length === EXPIRE_DATE_LENGTH &&
+      securityCode.length === SECURITY_CODE_LENGTH &&
+      password.join("").length === PASSWORD_LENGTH
+    );
+  }, [cardNumber, expireDate, securityCode, password]);
+
+  useEffect(() => {
+    if (isValidCardInfo()) {
+      setCanProceed(true);
+    } else {
+      setCanProceed(false);
+    }
+  }, [isValidCardInfo]);
```

Comment:
```suggestion
  const isValidCardInfo = useMemo(() => {
    const {
      NUMBER_UNIT_COUNT,
      NUMBER_UNIT_LENGTH,
      EXPIRE_DATE_LENGTH,
      SECURITY_CODE_LENGTH,
      PASSWORD_LENGTH,
    } = CARD_INFO_RULES;

    return (
      cardNumber.join("").length === NUMBER_UNIT_COUNT * NUMBER_UNIT_LENGTH &&
      expireDate.join("").length === EXPIRE_DATE_LENGTH &&
      securityCode.length === SECURITY_CODE_LENGTH &&
      password.join("").length === PASSWORD_LENGTH
    );
  }, [cardNumber, expireDate, securityCode, password]);

```
이렇게 변경해 볼 수도 있을 것 같아요.
`{isValidCardInfo && <Button text="다음" onClick={handleCardInfoSubmit} />}`

Reply:
말씀하신대로 useMemo를 사용하는 게 더 적합하겠네요! 
제안해주신 방식을 참고하면서 페어들과 같이 고민해봤는데요. 여기서 의존성을 더 나누어서 useMemo를 나누어 적용해주면 더 좋을 것 같다는 이야기가 나왔어요. 그래서 state에 따라 나누어 useMemo를 적용해봤습니다! [`9a1b248`](https://github.com/woowacourse/react-payments/pull/78/commits/9a1b248d54263eb3a9c258adc935fc406cad6f41)